### PR TITLE
Let kubectl api-versions use the discovery client

### PR DESCRIPTION
**Pull request by caesarxuchao**
The first commit is #15659.

The second commit fixes `kubectl api-versions`.

example output: [updated]

```
$ kubectl api-versions
extensions/v1beta1
v1
```

@kubernetes/goog-ux @lavalamp

### CODE REVIEW

**Comment by JanetKuo** [comment]
The code looks good. I have some further questions:
- Should we include group/version information in the help document?
- #15874 is planning to change group/type/name to type.group/name. Does this imply that we should show `Available Server Api Versions: [v1beta1.extensions v1]` instead? cc @smarterclayton @bgrant0607 @lavalamp

**Comment by caesarxuchao** [comment]
@JanetKuo seems we are not changing the "group/version" format, may I get the LGTM?

> Should we include group/version information in the help document?

which document are you referring to?

**Comment by JanetKuo** [comment]
@caesarxuchao I meant the .md file.

File: pkg/kubectl/cmd/apiversions.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ func NewCmdApiVersions(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 		Use: "api-versions",
 		// apiversions is deprecated.
 		Aliases: []string{"apiversions"},
-		Short:   "Print available API versions.",
+		Short:   "Print the supported API versions on the server, in the form of \"group/version\".",
```
Comment (caesarxuchao) [inline pkg/kubectl/cmd/apiversions.go]: @JanetKuo, I changed the description, hopefully it's clear.

**Comment by JanetKuo** [comment]
please update generated docs

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl

File: pkg/kubectl/cmd/apiversions.go [hunk 2]
```diff
@@ -51,19 +51,13 @@ func RunApiVersions(f *cmdutil.Factory, w io.Writer) error {
 		return err
 	}
 
-	apiVersions, err := client.ServerAPIVersions()
+	groupList, err := client.Discovery().ServerGroups()
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/apiversions.go]: What about 1.1 kubectl vs. 1.0 apiserver?

**Comment by bgrant0607** [comment]
cc @jlowdermilk @deads2k

File: pkg/kubectl/cmd/apiversions.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/apiversions.go]: Good point. We need to fix the ServerGroups to not return error if the apiserver is 1.0.

File: pkg/kubectl/cmd/apiversions.go [hunk 3]
```diff
@@ -51,19 +51,13 @@ func RunApiVersions(f *cmdutil.Factory, w io.Writer) error {
...
 
-	var expAPIVersions *unversioned.APIVersions
-	expAPIVersions, err = client.Extensions().ServerAPIVersions()
-
-	fmt.Fprintf(w, "Available Server Api Versions: %#v\n", *apiVersions)
-	if err == nil {
-		fmt.Fprintf(w, "Available Server Experimental Api Versions: %#v\n", *expAPIVersions)
-	}
-
+	fmt.Fprintf(w, "Available Server API Versions: %v\n", apiVersions)
```
Comment (deads2k) [inline pkg/kubectl/cmd/apiversions.go]: For human readable output, I think they'd prefer the strings joined or on separate lines and there should be a guaranteed sort.  The server side may be guaranteeing an order, but its not obvious from the client docs.

File: pkg/kubectl/cmd/apiversions.go [hunk 4]
```diff
@@ -51,19 +51,13 @@ func RunApiVersions(f *cmdutil.Factory, w io.Writer) error {
 		return err
 	}
 
-	apiVersions, err := client.ServerAPIVersions()
+	groupList, err := client.Discovery().ServerGroups()
 	if err != nil {
 		fmt.Printf("Couldn't get available api versions from server: %v\n", err)
```
Comment (deads2k) [inline pkg/kubectl/cmd/apiversions.go]: Either to `w` or `err`.

Comment (deads2k) [inline pkg/kubectl/cmd/apiversions.go]: Actually, It seems like this should return the `err` to later be checked. Having the exit here makes it impossible to reasonable test this path.

File: pkg/kubectl/cmd/apiversions.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/kubectl/cmd/apiversions.go]: The client should fall back to 1.0 behavior.

File: pkg/kubectl/cmd/apiversions.go (hunk 4 above)
Comment (bgrant0607) [inline pkg/kubectl/cmd/apiversions.go]: +1.

File: pkg/kubectl/cmd/apiversions.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/apiversions.go]: it's fixed in #16066

File: pkg/kubectl/cmd/apiversions.go (hunk 4 above)
Comment (bgrant0607) [inline pkg/kubectl/cmd/apiversions.go]: In general, business logic should return errors and neither output nor exit.

Additionally, normal output of commands should go to `out`, while error output should go to stderr.

**Comment by bgrant0607** [comment]
Do we need this for 1.1?

File: pkg/kubectl/cmd/apiversions.go (hunk 1 above)
Comment (JanetKuo) [inline pkg/kubectl/cmd/apiversions.go]: It is. Thanks.

**Comment by caesarxuchao** [comment]
Yes. Currently `kubectl api-versions` only returns v1, no matter if extensions is enabled on the server.

File: pkg/kubectl/cmd/apiversions.go [hunk 5]
```diff
@@ -51,19 +52,16 @@ func RunApiVersions(f *cmdutil.Factory, w io.Writer) error {
...
 
-	apiVersions, err := client.ServerAPIVersions()
+	groupList, err := client.Discovery().ServerGroups()
 	if err != nil {
-		fmt.Printf("Couldn't get available api versions from server: %v\n", err)
-		os.Exit(1)
+		return fmt.Errorf("Couldn't get available api versions from server: %v\n", err)
 	}
+	apiVersions := unversioned_client.ExtractGroupVersions(groupList)
+	sort.Strings(apiVersions)
```
Comment (caesarxuchao) [inline pkg/kubectl/cmd/apiversions.go]: add sort here.

File: pkg/kubectl/cmd/apiversions.go [hunk 6]
```diff
@@ -51,19 +52,16 @@ func RunApiVersions(f *cmdutil.Factory, w io.Writer) error {
 		return err
 	}
 
-	apiVersions, err := client.ServerAPIVersions()
+	groupList, err := client.Discovery().ServerGroups()
 	if err != nil {
-		fmt.Printf("Couldn't get available api versions from server: %v\n", err)
-		os.Exit(1)
+		return fmt.Errorf("Couldn't get available api versions from server: %v\n", err)
```
Comment (caesarxuchao) [inline pkg/kubectl/cmd/apiversions.go]: return error here.

**Comment by caesarxuchao** [comment]
example output:
```
$ kubectl api-versions
Available Server API Versions:
extensions/v1beta1
v1
```

File: pkg/kubectl/cmd/apiversions.go [hunk 7]
```diff
@@ -51,19 +52,16 @@ func RunApiVersions(f *cmdutil.Factory, w io.Writer) error {
...
 
-	var expAPIVersions *unversioned.APIVersions
-	expAPIVersions, err = client.Extensions().ServerAPIVersions()
-
-	fmt.Fprintf(w, "Available Server Api Versions: %#v\n", *apiVersions)
-	if err == nil {
-		fmt.Fprintf(w, "Available Server Experimental Api Versions: %#v\n", *expAPIVersions)
+	fmt.Fprintf(w, "Available Server API Versions:\n")
+	for _, v := range apiVersions {
+		fmt.Fprintf(w, "%v\n", v)
```
Comment (JanetKuo) [inline pkg/kubectl/cmd/apiversions.go]: `Fprintln`?

Comment (caesarxuchao) [inline pkg/kubectl/cmd/apiversions.go]: Thanks. Fixed.

**Comment by caesarxuchao** [comment]
We should add a test for `kubectl api-versions` in test-cmd.sh

File: pkg/kubectl/cmd/apiversions.go [hunk 8]
```diff
@@ -51,19 +52,16 @@ func RunApiVersions(f *cmdutil.Factory, w io.Writer) error {
...
+	apiVersions := unversioned_client.ExtractGroupVersions(groupList)
+	sort.Strings(apiVersions)
 
-	var expAPIVersions *unversioned.APIVersions
-	expAPIVersions, err = client.Extensions().ServerAPIVersions()
-
-	fmt.Fprintf(w, "Available Server Api Versions: %#v\n", *apiVersions)
-	if err == nil {
-		fmt.Fprintf(w, "Available Server Experimental Api Versions: %#v\n", *expAPIVersions)
+	fmt.Fprintf(w, "Available Server API Versions:\n")
```
Comment (deads2k) [inline pkg/kubectl/cmd/apiversions.go]: Optional: I don't see a need for this line.  What else would you be outputing?

**Comment by deads2k** [comment]
Optional nit.

lgtm.

File: pkg/kubectl/cmd/apiversions.go (hunk 8 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/apiversions.go]: Yes. Let's keep it simple.

**Comment by caesarxuchao** [comment]
Squashed and addressed the optional nit. Apply the lgtm label per @deads2k's comment.

**Comment by caesarxuchao** [comment]
e2e is a real one:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:436
Oct 22 10:03:54.154: Missing caption in kubectl api-versions
```

File: test/e2e/kubectl.go [hunk 9]
```diff
@@ -427,9 +427,6 @@ var _ = Describe("Kubectl client", func() {
 		It("should check if v1 is in available api versions [Conformance]", func() {
 			By("validating api verions")
 			output := runKubectl("api-versions")
-			if !strings.Contains(output, "Available Server Api Versions:") {
```
Comment (caesarxuchao) [inline test/e2e/kubectl.go]: Remove this check as I removed the caption

Comment (JanetKuo) [inline test/e2e/kubectl.go]: Please update the PR description too.

Comment (caesarxuchao) [inline test/e2e/kubectl.go]: Updated the example output. Thanks.

**Comment by caesarxuchao** [comment]
@k8s-bot test this please

**Comment by caesarxuchao** [comment]
@k8s-bot test this please

**Pull request merged**